packages/calcite-components/src/components/input/common/input.ts

```diff
@@ -0,0 +1,62 @@
+export type InputComponent = NumericInputComponent | TextualInputComponent;
+
+export interface NumericInputComponent {
+  min: number;
+  max: number;
+  step: number | "any";
+}
+
+export interface TextualInputComponent {
+  pattern?: string;
+  minLength: number;
+  maxLength: number;
+}
+
+const minMaxStepTypes = ["date", "datetime-local", "month", "number", "range", "time", "week"];
+const patternTypes = ["email", "password", "search", "tel", "text", "url"];
+const minMaxLengthTypes = ["email", "password", "search", "tel", "text", "textarea", "url"];
+
+function toString(num: number): string {
+  return num?.toString() ?? "";
+}
+
+/**
+ * Synchronizes the hidden form input with the validation-related input properties.
+ *
+ * @param type - The input type.
+ * @param inputComponent
+ * @param hiddenFormInput
+ */
+export function syncHiddenFormInput(
+  type: HTMLInputElement["type"] | "textarea",
+  inputComponent: InputComponent,
+  hiddenFormInput: HTMLInputElement,
+): void {
+  hiddenFormInput.type = type === "textarea" ? "text" : type;
+
+  if (minMaxStepTypes.includes(type)) {
+    const numericInputComponent = inputComponent as NumericInputComponent;
+    hiddenFormInput.min = toString(numericInputComponent.min);
+    hiddenFormInput.max = toString(numericInputComponent.max);
```

Comment:
I mentioned this on Teams, but adding it here for posterity - I think we will lose the extra precision from BigDecimal when validating the `min`/`max` constraints. I'm guessing the string is cast to a number internally when making the comparison. For example, I'm guessing the form will treat this as valid when it shouldn't:

```html
<calcite-input-number value="1.123456789123456789" max="1.123456789123456788">
```

```js
console.log(1.123456789123456789 > 1.123456789123456788) // => false
```

I don't know if we should care about that until someone logs an issue, but we should at least add a code comment for future reference.

Reply:
The potential solution I had in mind was adding custom logic in our `reportValidity()` method, which uses BigDecimal to make the comparison. We will likely need to do the same for other components like input-date-picker and input-time-picker anyway. 

We would need to call `reportValidity()` on all the form components with custom validation logic when submitting the form. If our logic finds invalid values, we'd call `setValidity()` on the hidden inputs to prevent form submission.

Reply:
I dig the beefed-up validation logic. 💪

I added a note regarding expected loss of precision in some cases based on your comment.